pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,361 @@
+	// - unknown output values (similar to computed)
+	// - nulls (they are allowed for any type, and for missing required properties see validatePropertyMap)
+	if propertyValue.IsResourceReference() ||
+		propertyValue.IsAsset() ||
+		propertyValue.IsArchive() ||
+		propertyValue.IsComputed() ||
+		propertyValue.IsNull() ||
+		(propertyValue.IsOutput() && !propertyValue.OutputValue().Known) {
+		return nil
+	}
+
+	// for known first-class outputs simply validate their known value
+	if propertyValue.IsOutput() && propertyValue.OutputValue().Known {
+		elementValue := propertyValue.OutputValue().Element
+		return v.validatePropertyValue(elementValue, typeSpec, propertyPath)
+	}
+
+	// for secrets validate their inner value
+	if propertyValue.IsSecret() {
+		elementValue := propertyValue.SecretValue().Element
+		return v.validatePropertyValue(elementValue, typeSpec, propertyPath)
+	}
+
+	// now we are going to switch on the desired type
+	//
+	// a good reference here for the semantics of the TypeSpec is
+	// https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/bind.go#L881
+	//
+	// this code is not using the binder functionality as that would require a plugin loader to resolve references,
+	// which is not desirable for the provider runtime
+	//
+	// possibly this could be done in the future with a way to still use the bind code while resolving references to
+	// unchecked types as using the bound representation would make the code cleaner
+	//
+	// for now the code simply follows bindSpec
+
+	if typeSpec.Ref != "" {
+		objType := v.getType(typeSpec.Ref)
+		// Refusing to validate unknown or unresolved type.
+		if objType == nil {
+			return nil
+		}
+
+		if !propertyValue.IsObject() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected object type, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+
+		return v.validatePropertyMap(
+			propertyValue.ObjectValue(),
+			objType.Required,
+			objType.Properties,
+			propertyPath,
+		)
+	}
+
+	if typeSpec.OneOf != nil {
+		// bindTypeSpecOneOf provides a good hint of how to interpret these:
+		//
+		// https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/bind.go#L842
+		//
+		// Specifically it defines the defaultType, discriminator, mapping and elements.
+		if len(typeSpec.OneOf) < 2 {
+			// invalid type, don't validate
+			return nil
+		}
+
+		// don't validate types with Discriminator (yet)
+		if typeSpec.Discriminator != nil {
+			return nil
+		}
+
+		// default type
+		if typeSpec.Type != "" {
+			typeSpec.OneOf = append(typeSpec.OneOf, pschema.TypeSpec{Type: typeSpec.Type})
+		}
+
+		failures := []TypeFailure{}
+		for _, spec := range typeSpec.OneOf {
+			failure := v.validatePropertyValue(propertyValue, spec, propertyPath)
+			if len(failure) == 0 {
+				// we have found the correct type
+				return nil
+			}
+			failures = append(failures, failure...)
```

Comment:
Merging all these failures will be unreadable. Perhaps just ignore them all for now and emit one generic failure that "matched none of the union type members"

Reply:
I took a stab at trying to return a better single error message, let me know what you think.